Source/WebCore/rendering/svg/RenderSVGImage.cpp

```diff
@@ -0,0 +1,413 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGImage.h"
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+#include "AXObjectCache.h"
+#include "BitmapImage.h"
+#include "DocumentInlines.h"
+#include "GeometryUtilities.h"
+#include "GraphicsContext.h"
+#include "HitTestResult.h"
+#include "LayoutRepainter.h"
+#include "PointerEventsHitRules.h"
+#include "RenderImageResource.h"
+#include "RenderLayer.h"
+#include "RenderSVGResource.h"
+#include "RenderSVGResourceFilter.h"
+#include "SVGElementTypeHelpers.h"
+#include "SVGImageElement.h"
+#include "SVGRenderingContext.h"
+#include "SVGResources.h"
+#include "SVGResourcesCache.h"
+#include <wtf/IsoMallocInlines.h>
+#include <wtf/StackStats.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGImage);
+
+RenderSVGImage::RenderSVGImage(SVGImageElement& element, RenderStyle&& style)
+    : RenderSVGModelObject(element, WTFMove(style))
+    , m_imageResource(makeUnique<RenderImageResource>())
+{
+    imageResource().initialize(*this);
+}
+
+RenderSVGImage::~RenderSVGImage() = default;
+
+void RenderSVGImage::willBeDestroyed()
+{
+    imageResource().shutdown();
+    RenderSVGModelObject::willBeDestroyed();
+}
+
+SVGImageElement& RenderSVGImage::imageElement() const
+{
+    return downcast<SVGImageElement>(RenderSVGModelObject::element());
+}
+
+FloatRect RenderSVGImage::calculateObjectBoundingBox() const
+{
+    LayoutSize intrinsicSize;
+    if (CachedImage* cachedImage = imageResource().cachedImage())
+        intrinsicSize = cachedImage->imageSizeForRenderer(nullptr, style().effectiveZoom());
+
+    SVGLengthContext lengthContext(&imageElement());
+
+    Length width = style().width();
+    Length height = style().height();
+
+    float concreteWidth;
+    if (!width.isAuto())
+        concreteWidth = lengthContext.valueForLength(width, SVGLengthMode::Width);
+    else if (!height.isAuto() && !intrinsicSize.isEmpty())
+        concreteWidth = lengthContext.valueForLength(height, SVGLengthMode::Height) * intrinsicSize.width() / intrinsicSize.height();
+    else
+        concreteWidth = intrinsicSize.width();
+
+    float concreteHeight;
+    if (!height.isAuto())
+        concreteHeight = lengthContext.valueForLength(height, SVGLengthMode::Height);
+    else if (!width.isAuto() && !intrinsicSize.isEmpty())
+        concreteHeight = lengthContext.valueForLength(width, SVGLengthMode::Width) * intrinsicSize.height() / intrinsicSize.width();
+    else
+        concreteHeight = intrinsicSize.height();
```

Comment:
These two blocks feel like they can be done using one lambda, but maybe it is hard to do here.

Reply:
There will be a cleanup patch anyhow for the other RenderSVG* files as well - let's postpone cleanups (the code is copied 1:1 from LegacyRenderSVGImage)